// use-local-llm/dist/index.js
'use strict';

var react = require('react');

// src/hooks/useLocalLLM.ts

// src/utils/endpoints.ts
var ENDPOINTS = {
  ollama: {
    url: "http://localhost:11434",
    backend: "ollama"
  },
  lmstudio: {
    url: "http://localhost:1234",
    backend: "lmstudio"
  },
  llamacpp: {
    url: "http://localhost:8080",
    backend: "llamacpp"
  }
};
var CHAT_PATHS = {
  ollama: "/api/chat",
  lmstudio: "/v1/chat/completions",
  llamacpp: "/v1/chat/completions",
  "openai-compatible": "/v1/chat/completions"
};
var GENERATE_PATHS = {
  ollama: "/api/generate",
  lmstudio: "/v1/completions",
  llamacpp: "/v1/completions",
  "openai-compatible": "/v1/completions"
};
var MODEL_LIST_PATHS = {
  ollama: "/api/tags",
  lmstudio: "/v1/models",
  llamacpp: "/v1/models",
  "openai-compatible": "/v1/models"
};
var PORT_TO_BACKEND = {
  "11434": "ollama",
  "1234": "lmstudio",
  "8080": "llamacpp"
};
function detectBackend(endpoint) {
  try {
    const url = new URL(endpoint);
    return PORT_TO_BACKEND[url.port] ?? "openai-compatible";
  } catch {
    return "openai-compatible";
  }
}

// src/utils/streamParser.ts
function parseStreamChunk(raw, backend) {
  const trimmed = raw.trim();
  if (!trimmed || trimmed === "data: [DONE]") {
    return trimmed === "data: [DONE]" ? { content: "", done: true } : null;
  }
  let jsonStr = trimmed;
  if (trimmed.startsWith("data: ")) {
    jsonStr = trimmed.slice(6);
  }
  try {
    const parsed = JSON.parse(jsonStr);
    if (backend === "ollama") {
      if (parsed.message) {
        return {
          content: parsed.message.content ?? "",
          done: !!parsed.done,
          model: parsed.model
        };
      }
      return {
        content: parsed.response ?? "",
        done: !!parsed.done,
        model: parsed.model
      };
    }
    const choice = parsed.choices?.[0];
    if (choice) {
      return {
        content: choice.delta?.content ?? choice.text ?? "",
        done: choice.finish_reason === "stop",
        model: parsed.model
      };
    }
    return null;
  } catch {
    return null;
  }
}
function buildChatBody(options, backend) {
  if (backend === "ollama") {
    return JSON.stringify({
      model: options.model,
      messages: options.messages,
      stream: true,
      ...options.temperature != null && {
        options: { temperature: options.temperature }
      }
    });
  }
  return JSON.stringify({
    model: options.model,
    messages: options.messages,
    stream: true,
    ...options.temperature != null && { temperature: options.temperature }
  });
}
function buildGenerateBody(options, backend) {
  if (backend === "ollama") {
    return JSON.stringify({
      model: options.model,
      prompt: options.prompt,
      stream: true,
      ...options.temperature != null && {
        options: { temperature: options.temperature }
      }
    });
  }
  return JSON.stringify({
    model: options.model,
    prompt: options.prompt,
    stream: true,
    ...options.temperature != null && { temperature: options.temperature }
  });
}
async function* streamChat(options) {
  const backend = options.backend ?? detectBackend(options.endpoint);
  const url = `${options.endpoint.replace(/\/$/, "")}${CHAT_PATHS[backend]}`;
  const response = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: buildChatBody(options, backend),
    signal: options.signal
  });
  if (!response.ok) {
    const text = await response.text().catch(() => "Unknown error");
    throw new Error(`LLM request failed (${response.status}): ${text}`);
  }
  yield* readStream(response, backend);
}
async function* streamGenerate(options) {
  const backend = options.backend ?? detectBackend(options.endpoint);
  const url = `${options.endpoint.replace(/\/$/, "")}${GENERATE_PATHS[backend]}`;
  const response = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: buildGenerateBody(options, backend),
    signal: options.signal
  });
  if (!response.ok) {
    const text = await response.text().catch(() => "Unknown error");
    throw new Error(`LLM request failed (${response.status}): ${text}`);
  }
  yield* readStream(response, backend);
}
async function* readStream(response, backend) {
  const reader = response.body?.getReader();
  if (!reader) {
    throw new Error("Response body is not readable");
  }
  const decoder = new TextDecoder();
  let buffer = "";
  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop() ?? "";
      for (const line of lines) {
        const chunk = parseStreamChunk(line, backend);
        if (chunk) {
          yield chunk;
          if (chunk.done) return;
        }
      }
    }
    if (buffer.trim()) {
      const chunk = parseStreamChunk(buffer, backend);
      if (chunk) yield chunk;
    }
  } finally {
    reader.releaseLock();
  }
}

// src/hooks/useLocalLLM.ts
function useLocalLLM(options) {
  const [messages, setMessages] = react.useState([]);
  const [isStreaming, setIsStreaming] = react.useState(false);
  const [isLoading, setIsLoading] = react.useState(false);
  const [error, setError] = react.useState(null);
  const abortControllerRef = react.useRef(null);
  const optionsRef = react.useRef(options);
  optionsRef.current = options;
  const messagesRef = react.useRef(messages);
  messagesRef.current = messages;
  const abort = react.useCallback(() => {
    abortControllerRef.current?.abort();
    abortControllerRef.current = null;
    setIsStreaming(false);
    setIsLoading(false);
  }, []);
  const clear = react.useCallback(() => {
    abort();
    setMessages([]);
    setError(null);
  }, [abort]);
  const send = react.useCallback(async (content) => {
    const opts = optionsRef.current;
    const controller = new AbortController();
    abortControllerRef.current?.abort();
    abortControllerRef.current = controller;
    setError(null);
    setIsLoading(true);
    const userMessage = { role: "user", content };
    const assistantMessage = { role: "assistant", content: "" };
    const updated = [...messagesRef.current, userMessage];
    const withAssistant = [...updated, assistantMessage];
    setMessages(withAssistant);
    const apiMessages = [];
    if (opts.system) {
      apiMessages.push({ role: "system", content: opts.system });
    }
    apiMessages.push(...updated);
    let accumulated = "";
    try {
      const backend = opts.backend ?? detectBackend(opts.endpoint);
      const stream = streamChat({
        endpoint: opts.endpoint,
        backend,
        model: opts.model,
        messages: apiMessages,
        temperature: opts.temperature,
        signal: controller.signal
      });
      setIsStreaming(true);
      setIsLoading(false);
      for await (const chunk of stream) {
        if (controller.signal.aborted) break;
        accumulated += chunk.content;
        opts.onToken?.(chunk.content);
        setMessages((prev) => {
          const next = [...prev];
          next[next.length - 1] = { role: "assistant", content: accumulated };
          return next;
        });
      }
      const finalMessage = { role: "assistant", content: accumulated };
      opts.onResponse?.(finalMessage);
    } catch (err) {
      if (err instanceof DOMException && err.name === "AbortError") {
        return;
      }
      const streamError = err instanceof Error ? err : new Error("Stream failed");
      setError(streamError);
      opts.onError?.(streamError);
    } finally {
      setIsStreaming(false);
      setIsLoading(false);
      abortControllerRef.current = null;
    }
  }, []);
  return { messages, send, isStreaming, isLoading, abort, error, clear };
}
function useModelList(options) {
  const [models, setModels] = react.useState([]);
  const [isLoading, setIsLoading] = react.useState(false);
  const [error, setError] = react.useState(null);
  const optionsRef = react.useRef(options);
  optionsRef.current = options;
  const fetchModels = react.useCallback(async () => {
    const opts = optionsRef.current;
    const endpoint = opts?.endpoint ?? ENDPOINTS.ollama.url;
    const backend = opts?.backend ?? detectBackend(endpoint);
    const url = `${endpoint.replace(/\/$/, "")}${MODEL_LIST_PATHS[backend]}`;
    setIsLoading(true);
    setError(null);
    try {
      const response = await fetch(url);
      if (!response.ok) {
        throw new Error(`Failed to fetch models (${response.status})`);
      }
      const data = await response.json();
      const rawModels = data.models ?? data.data ?? [];
      const parsed = rawModels.map((m) => {
        const model = m;
        return {
          name: model.name ?? model.id ?? "unknown",
          size: model.size,
          modifiedAt: model.modified_at ?? model.created,
          digest: model.digest
        };
      });
      setModels(parsed);
    } catch (err) {
      const fetchError = err instanceof Error ? err : new Error("Failed to fetch models");
      setError(fetchError);
    } finally {
      setIsLoading(false);
    }
  }, []);
  react.useEffect(() => {
    fetchModels();
  }, [fetchModels]);
  return { models, isLoading, error, refresh: fetchModels };
}

// src/hooks/useOllama.ts
function useOllama(model, options) {
  return useLocalLLM({
    endpoint: options?.endpoint ?? ENDPOINTS.ollama.url,
    backend: "ollama",
    model,
    system: options?.system,
    temperature: options?.temperature,
    onToken: options?.onToken,
    onResponse: options?.onResponse,
    onError: options?.onError
  });
}
function useStreamCompletion(options) {
  const [text, setText] = react.useState("");
  const [tokens, setTokens] = react.useState([]);
  const [isStreaming, setIsStreaming] = react.useState(false);
  const [error, setError] = react.useState(null);
  const abortControllerRef = react.useRef(null);
  const optionsRef = react.useRef(options);
  optionsRef.current = options;
  const abort = react.useCallback(() => {
    abortControllerRef.current?.abort();
    abortControllerRef.current = null;
    setIsStreaming(false);
  }, []);
  const start = react.useCallback(async () => {
    abort();
    const opts = optionsRef.current;
    const controller = new AbortController();
    abortControllerRef.current = controller;
    setText("");
    setTokens([]);
    setError(null);
    setIsStreaming(true);
    let accumulated = "";
    try {
      const backend = opts.backend ?? detectBackend(opts.endpoint);
      const stream = streamGenerate({
        endpoint: opts.endpoint,
        backend,
        model: opts.model,
        prompt: opts.prompt,
        temperature: opts.temperature,
        signal: controller.signal
      });
      for await (const chunk of stream) {
        if (controller.signal.aborted) break;
        accumulated += chunk.content;
        setText(accumulated);
        setTokens((prev) => [...prev, chunk.content]);
        opts.onToken?.(chunk.content);
      }
      opts.onComplete?.(accumulated);
    } catch (err) {
      if (err instanceof DOMException && err.name === "AbortError") {
        return;
      }
      const error2 = err instanceof Error ? err : new Error("Stream failed");
      setError(error2);
      opts.onError?.(error2);
    } finally {
      setIsStreaming(false);
      abortControllerRef.current = null;
    }
  }, [abort]);
  return { text, tokens, isStreaming, start, abort, error };
}

exports.CHAT_PATHS = CHAT_PATHS;
exports.ENDPOINTS = ENDPOINTS;
exports.GENERATE_PATHS = GENERATE_PATHS;
exports.MODEL_LIST_PATHS = MODEL_LIST_PATHS;
exports.detectBackend = detectBackend;
exports.parseStreamChunk = parseStreamChunk;
exports.streamChat = streamChat;
exports.streamGenerate = streamGenerate;
exports.useLocalLLM = useLocalLLM;
exports.useModelList = useModelList;
exports.useOllama = useOllama;
exports.useStreamCompletion = useStreamCompletion;
